### PR TITLE
[GLUTEN-3057][VL] Crash by dangling pointer to Arrow memory pool in shuffle writer, when task is killed

### PR DESCRIPTION
**Pull request by zhztheplayer**

### CODE REVIEW

File: cpp/core/shuffle/ShuffleWriter.h [hunk 1]
```diff
@@ -234,15 +234,17 @@ class ShuffleWriter {
       : numPartitions_(numPartitions),
         partitionWriterCreator_(std::move(partitionWriterCreator)),
         options_(std::move(options)),
-        codec_(createArrowIpcCodec(options_.compression_type, options_.codec_backend)),
-        pool_(std::make_shared<ShuffleBufferPool>(options_.memory_pool)) {}
+        pool_(std::make_shared<ShuffleBufferPool>(options_.memory_pool)),
+        codec_(createArrowIpcCodec(options_.compression_type, options_.codec_backend)) {}
```
Comment (zhouyuan) [inline cpp/core/shuffle/ShuffleWriter.h:238]: so the fix is to move this line up?

Comment (zhztheplayer) [reply cpp/core/shuffle/ShuffleWriter.h:238]: Yes, the poll should be alive during the buffers are destructed,

**Comment by zhouyuan** [comment]
@jackylee-ch

**Review by zhouyuan — APPROVED** [review]
👍

**Pull request merged**